### PR TITLE
Use address sanitizer for core worker

### PR DESCRIPTION
**Pull request by pcmoritz**
## Why are these changes needed?

This PR makes it so the core worker test is run with the address sanitizer.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/test/run_core_worker_tests.sh [hunk 1]
```diff
@@ -6,7 +6,7 @@
 set -e
 set -x
 
-bazel build "//:core_worker_test" "//:mock_worker"  "//:raylet" "//:libray_redis_module.so" "@plasma//:plasma_store_server"
+bazel build -c dbg --config=asan "//:core_worker_test" "//:mock_worker" "//:raylet" "//:libray_redis_module.so" "@plasma//:plasma_store_server"
```
Comment (zhijunfu) [inline src/ray/test/run_core_worker_tests.sh]: running debug build might be very slow, is this required by asan?

File: src/ray/common/id.cc [hunk 2]
```diff
@@ -154,49 +154,49 @@ uint8_t ObjectID::GetTransportType() const {
 
 // This code is from https://sites.google.com/site/murmurhash/
 // and is public domain.
-uint64_t MurmurHash64A(const void *key, int len, unsigned int seed) {
-  const uint64_t m = 0xc6a4a7935bd1e995;
-  const int r = 47;
+uint64_t MurmurHash64B (const void * key, int len, uint64_t seed) {
```
Comment (zhijunfu) [inline src/ray/common/id.cc]: is this change related to asan?

**Comment by zhijunfu** [comment]
@pcmoritz Should we consider to enable memory and thread sanitizer as well?

File: .bazelrc [hunk 3]
```diff
@@ -29,3 +29,37 @@ build --http_timeout_scaling=5.0
...
+# Address sanitizer configuration:
+# Unfortunately using the function and vptr sanitizers of the
+# undefined sanitizer produces linker errors, so they are switched
+# off for now (see also https://github.com/google/oss-fuzz/issues/713)
+build:asan --strip=never
+build:asan --copt -fsanitize=address
+build:asan --copt -fsanitize=undefined
+build:asan --copt -fno-sanitize=function,vptr
+build:asan --copt -DADDRESS_SANITIZER
+build:asan --copt -O2
```
Comment (mehrdadn) [inline .bazelrc]: Can we remove the flags not related to ASAN?

File: .bazelrc [hunk 4]
```diff
@@ -29,3 +29,37 @@ build --http_timeout_scaling=5.0
...
+# Thread sanitizer configuration:
+# Unfortunately using the function and vptr sanitizers of the
+# undefined sanitizer produces linker errors, so they are switched
+# off for now (see also https://github.com/google/oss-fuzz/issues/713)
+build:tsan --strip=never
+build:tsan --copt -fsanitize=thread
+build:tsan --copt -fsanitize=undefined
+build:tsan --copt -fno-sanitize=function,vptr
+build:tsan --copt -DTHREAD_SANITIZER
+build:tsan --copt -O2
```
Comment (mehrdadn) [inline .bazelrc]: Ditto above

**Review by mehrdadn — COMMENTED** [review]
Instead of modifying `.bazelrc`, can we modify `COPTS` flags in the relevant build files? You can use conditionals there, and this seems like a perfect use case for it.  
`.bazelrc` really seems like a last resort for modifying build flags; we really shouldn't proliferate flags in it if we can avoid it. It becomes kind of a mess to keep maintained and it isn't modular across projects (not to mention it will make it harder for others to depend on us too).

File: src/ray/common/id.cc [hunk 5]
```diff
@@ -20,7 +20,7 @@ extern "C" {
 
 namespace ray {
 
-uint64_t MurmurHash64A(const void *key, int len, unsigned int seed);
+uint64_t MurmurHash64B(const void *key, int len, uint64_t seed);
```
Comment (mehrdadn) [inline src/ray/common/id.cc]: The `int` should also be `size_t`.

File: src/ray/common/id.cc [hunk 6]
```diff
@@ -172,49 +172,65 @@ uint8_t ObjectID::GetTransportType() const {
...
-  case 6:
-    h ^= uint64_t(data2[5]) << 40;
-  case 5:
-    h ^= uint64_t(data2[4]) << 32;
-  case 4:
-    h ^= uint64_t(data2[3]) << 24;
+  switch (len) {
   case 3:
-    h ^= uint64_t(data2[2]) << 16;
+    h2 ^= ((unsigned char *)data)[2] << 16;
```
Comment (mehrdadn) [inline src/ray/common/id.cc]: Can we use a C++ cast here instead? I realize it's longer, but it's less error-prone. I think `static_cast` should suffice here. (Also, `const` shouldn't be cast away.)

File: .travis.yml [hunk 7]
```diff
@@ -80,6 +80,31 @@ matrix:
...
+        - export PATH="$HOME/miniconda/bin:$PATH"
+        - ./ci/suppress_output ./ci/travis/install-ray.sh
+
+        # Run raylet integration tests with address sanitizer
+        - export RAY_BAZEL_CONFIG=--config=asan
+        - ./ci/suppress_output bash src/ray/test/run_core_worker_tests.sh
+        - ./ci/suppress_output bash src/ray/test/run_object_manager_tests.sh
+
+        # Run raylet integration tests with thread sanitizer
+        - export RAY_BAZEL_CONFIG=--config=tsan
```
Comment (mehrdadn) [inline .travis.yml]: This is going to leak one of the `RAY_BAZEL_CONFIG` across the subsequent statements, right? If it needs to be an environment variable, it would be better to scope it to them if possible. Maybe like:

	(
		export RAY_BAZEL_CONFIG=...
		...
		...
	)

But maybe that's not necessary? Can we just pass these as arguments to be forwarded to Bazel inside each script (via `"$@"`)? It'd be nice to keep the arrays intact (the space splitting of `$RAY_BAZEL_CONFIG` makes me uncomfortable).

File: src/ray/test/run_core_worker_tests.sh (hunk 1 above)
Comment (mehrdadn) [reply src/ray/test/run_core_worker_tests.sh]: Do debug builds work at all (given the `DEBUG` issue)?

File: .travis.yml (hunk 7 above)
Comment (pcmoritz) [reply .travis.yml]: Thanks, I fixed it a bit differently in https://github.com/ray-project/ray/pull/6701 which replaces this PR (instead of making it a global variable, I make it a local one). Not sure what you mean with space splitting.

File: .bazelrc (hunk 4 above)
Comment (pcmoritz) [reply .bazelrc]: Yeah, I just kept the flags needed for thread sanitizer in https://github.com/ray-project/ray/pull/6701.

File: src/ray/common/id.cc (hunk 5 above)
Comment (pcmoritz) [reply src/ray/common/id.cc]: This code is copied from murmurhash, but I removed it in https://github.com/ray-project/ray/pull/6701 (we might need to bring it back thought.

File: .travis.yml (hunk 7 above)
Comment (mehrdadn) [reply .travis.yml]: Oh I was just referring to the fact that missing quotes cause splitting on spaces, which is generally bad.

**Comment by mehrdadn** [comment]
@pcmoritz Is this PR still relevant, or was it superceded?

**Comment by pcmoritz** [comment]
@mehrdadn The thread sanitizer part got merged, the address sanitizer part not yet.

**Comment by robertnishihara** [comment]
@pcmoritz should this be closed?

**Comment by rkooo567** [comment]
Close because we now have ASAN build.
